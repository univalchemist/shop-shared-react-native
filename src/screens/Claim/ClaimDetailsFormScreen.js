import {
  Box,
  Button,
  Container,
  Datepicker,
  ScrollView,
  StepProgressBar,
  Text,
} from '@wrappers/components';
import {
  CheckBoxField,
  InputField,
  SelectField,
} from '@wrappers/components/form';
import { normalizeAmount } from '@wrappers/core/normalizers';
import {
  validateConsultationDate,
  validateClaimType,
  validateClaimReason,
  validateDiagnosisText,
  hasValue,
} from '@wrappers/core/validations';
import {
  CLAIM_TYPE_MODAL,
  CLAIM_REASON_MODAL,
  CLAIM_UPLOAD_DOCUMENTS,
} from '@routes';
import moment from 'moment/min/moment-with-locales';
import PropTypes from 'prop-types';
import React from 'react';
import { injectIntl, FormattedMessage } from 'react-intl';
import { Icon } from 'react-native-elements';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { reduxForm } from 'redux-form';
import styled from 'styled-components/native';
import WalletBalancePanel, {
  WalletBalancePanelError,
  WalletBalanceSkeletonPlaceholder,
} from '@components/WalletBalancePanel';
import { useFetchActions, useTheme } from '@wrappers/core/hooks';
import { fetchWallet } from '@store/wallet/actions';
import { CLAIM_REASON_OTHERS } from './constants';
import { categories, logAction } from '@store/analytics/trackingActions';

const PrefixLeftText = styled(Text)`
  font-size: 16px;
  font-weight: 300;
  ${({ theme }) => `
    color: ${theme.inputField.rightText}
  `};
`;

const styles = {
  infoWrapper: {
    flexDirection: 'row',
  },
  infoText: {
    flex: 1,
  },
  checkboxField: {
    marginLeft: 0,
    padding: 0,
  },
};

const daysAgo = days => {
  const today = new Date();
  const priorDate = new Date().setDate(today.getDate() - days);
  return new Date(priorDate);
};

const ClaimDetailsFormScreen = ({
  claimType,
  fetchWallet,
  handleSubmit,
  change,
  claimTypeId,
  claimReason,
  touch,
  isMultiInsurer,
  intl,
  availableBalance,
  isMaternity = false,
  terminationDate,
  navigation: { navigate },
}) => {
  const theme = useTheme();
  const { types } = claimType;
  const selectedClaimType = types.byId[claimTypeId];
  const selectedClaimReason = claimType?.reasons?.byId[claimReason]?.code;
  const { maxAmountPerClaim: maxReceiptAmount, referralRequired } =
    selectedClaimType || {};

  const isInsuranceClaim = selectedClaimType?.isInsuranceClaim ?? true;
  let showWalletBallance = false;
  if (
    selectedClaimType &&
    claimType.categories.byId[selectedClaimType.claimCategoryId] &&
    claimType.categories.byId[
      selectedClaimType.claimCategoryId
    ].code.toLowerCase() === 'wellness'
  )
    // Only show wallet for WELLNESS
    showWalletBallance = true;
  const [isWalletLoading, isWalletError] = useFetchActions(
    [fetchWallet],
    showWalletBallance,
  );

  const intlMsg = (field, options) =>
    intl.formatMessage({ id: `claim.${field}` }, options);

  const maxReceiptMoney = intl.formatNumber(maxReceiptAmount, {
    format: 'money',
  });

  const receiptAmountHint = maxReceiptAmount
    ? intlMsg('label.hint.receiptAmount', {
        maxReceiptAmount: maxReceiptMoney,
      })
    : undefined;

  const validateMaxAmount = value => {
    if (
      typeof maxReceiptAmount === 'number' &&
      Number(value) &&
      Number(value) > maxReceiptAmount
    ) {
      const formattedAmount = intl.formatNumber(maxReceiptAmount, {
        format: 'money',
      });
      return intl.formatMessage(
        { id: 'exceedsMaximumAmount' },
        { amount: formattedAmount },
      );
    }
    return '';
  };

  const validateMoneyAmount = value => {
    if (!hasValue(value))
      return intl.formatMessage({ id: 'receiptAmountRequired' });
    else if (Number.isNaN(Number(value)))
      return intl.formatMessage({ id: 'moneyAmountInvalid' });
    else if (value > 0) return '';
    else return intl.formatMessage({ id: 'moneyAmountGreaterThanZero' });
  };

  const validateOtherInsurerAmount = (value, allValues) => {
    if (!allValues.isMultiInsurer) return '';
    if (!hasValue(value))
      return intl.formatMessage({ id: 'otherInsurerAmountRequired' });
    return Number(value) < Number(allValues.receiptAmount)
      ? ''
      : intl.formatMessage({ id: 'otherInsurerAmountMustBeLower' });
  };

  const maxDate = () => {
    return terminationDate &&
      new Date(terminationDate).getTime() < new Date().getTime()
      ? moment
          .utc(terminationDate)
          .local()
          .toDate()
      : new Date();
  };
  return (
    <>
      <StepProgressBar currentStep={2} stepCount={4} />
      {showWalletBallance &&
        (isWalletLoading ? (
          <WalletBalanceSkeletonPlaceholder />
        ) : isWalletError ? (
          <WalletBalancePanelError />
        ) : availableBalance === undefined ? (
          <WalletBalancePanelError />
        ) : (
          <WalletBalancePanel balance={availableBalance} />
        ))}

      <ScrollView>
        <Container>
          {showWalletBallance && availableBalance === 0 ? (
            <Box mb={32}>
              <Text color="error.0">
                <FormattedMessage id="claim.wellness.balance.warning" />
              </Text>
            </Box>
          ) : null}
          <Box mt={2}>
            <Box>
              <Datepicker
                defaultDate={maxDate()}
                maximumDate={maxDate()}
                minimumDate={daysAgo(90)}
                onConfirm={date => {
                  change('consultationDate', moment(date).toISOString());
                  logAction({
                    category: categories.CLAIMS_SUBMISSION,
                    action: 'Select consultation date',
                  });
                }}
                onCancel={() => touch('consultationDate')}
                field={({ setIsVisible }) => (
                  <SelectField
                    name="consultationDate"
                    validate={validateConsultationDate}
                    format={date => date && moment(date).format('ll')}
                    label={intlMsg('label.consultationDate')}
                    hint={intlMsg('label.hint.consultationDate')}
                    onPress={() => setIsVisible(true)}
                    onRight={({ color }) => <Icon name="event" color={color} />}
                  />
                )}
              />
            </Box>
            <Box>
              <SelectField
                name="claimType"
                validate={validateClaimType}
                label={intlMsg('label.claimType')}
                onPress={() => navigate(CLAIM_TYPE_MODAL)}
                onRight={({ color }) => (
                  <Icon name="expand-more" color={color} />
                )}
              />
            </Box>
            <Box>
              <SelectField
                name="claimReason"
                format={value =>
                  value && claimType.reasons.byId[value].claimReason
                }
                validate={validateClaimReason}
                disabled={!claimTypeId}
                label={intlMsg('label.claimReason')}
                onPress={() => navigate(CLAIM_REASON_MODAL)}
                onRight={({ color }) => (
                  <Icon name="expand-more" color={color} />
                )}
              />
            </Box>
            {selectedClaimReason === CLAIM_REASON_OTHERS ? (
              <Box>
                <InputField
                  name="diagnosisText"
                  validate={validateDiagnosisText}
                  maxLength={20}
                  returnKeyType="done"
                  label={intlMsg('label.claimOtherReason')}
                  customStyles={theme.customInputStyles}
                />
              </Box>
            ) : null}

            <Box>
              <InputField
                name="receiptAmount"
                validate={[validateMoneyAmount, validateMaxAmount]}
                validationErrorsLocalized
                keyboardType="numeric"
                returnKeyType="done"
                normalize={normalizeAmount}
                leftIcon={<PrefixLeftText>HK$ </PrefixLeftText>}
                label={
                  isInsuranceClaim
                    ? intlMsg('label.receiptAmount')
                    : intlMsg('label.claimAmount')
                }
                hint={receiptAmountHint}
                customStyles={theme.customInputStyles}
              />
            </Box>
            <Box mb={20}>
              <CheckBoxField
                name="isMaternity"
                size={24}
                containerStyle={styles.checkboxField}
                label={intlMsg('label.isMaternity')}
              />
              {isMaternity ? (
                <Box
                  p={10}
                  style={styles.infoWrapper}
                  backgroundColor={theme.colors.info[2]}
                >
                  <Icon name="info" color={theme.colors.blue[3]} />
                  <Text
                    ml={10}
                    style={styles.infoText}
                    color={theme.colors.gray[0]}
                  >
                    {intl.formatMessage({ id: 'claim.maternityNote' })}
                  </Text>
                </Box>
              ) : null}
            </Box>

            {isInsuranceClaim ? (
              <Box paddingBottom={34} paddingRight={10}>
                <CheckBoxField
                  name="isMultiInsurer"
                  size={24}
                  containerStyle={styles.checkboxField}
                  label={intlMsg('withAnotherInsurerText')}
                  onChange={data => {
                    if (!data) {
                      change('otherInsurerAmount', '');
                    } else {
                      logAction({
                        category: categories.CLAIMS_SUBMISSION,
                        action: 'Made a claim with another insurer',
                      });
                    }
                  }}
                />
              </Box>
            ) : null}

            {isInsuranceClaim && isMultiInsurer && (
              <Box>
                <InputField
                  name="otherInsurerAmount"
                  label={intlMsg('label.otherInsurerAmount')}
                  hint={intlMsg('label.hint.otherInsurerAmount')}
                  validationErrorsLocalized
                  validate={[validateMoneyAmount, validateOtherInsurerAmount]}
                  keyboardType="numeric"
                  returnKeyType="done"
                  normalize={normalizeAmount}
                  leftIcon={<PrefixLeftText>HK$ </PrefixLeftText>}
                  customStyles={theme.customInputStyles}
                />
              </Box>
            )}
            <Box>
              <Button
                primary
                onPress={handleSubmit(() => {
                  if (!referralRequired) {
                    change('referrals', []);
                  }
                  if (!isMultiInsurer) {
                    change('otherInsurerAmount', '');
                  }
                  navigate(CLAIM_UPLOAD_DOCUMENTS);
                })}
                title={intlMsg('uploadDocuments')}
                disabled={
                  showWalletBallance
                    ? isWalletLoading || availableBalance <= 0
                    : false
                }
              />
            </Box>
          </Box>
        </Container>
      </ScrollView>
    </>
  );
};

ClaimDetailsFormScreen.propTypes = {
  handleSubmit: PropTypes.func,
  theme: PropTypes.shape({}),
  claimType: PropTypes.shape({ selectedClaimItem: PropTypes.string }),
  claimReason: PropTypes.number,
  change: PropTypes.func.isRequired,
  claimTypeId: PropTypes.number,
  touch: PropTypes.func.isRequired,
  isMultiInsurer: PropTypes.bool,
  intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,
  maxReceiptAmount: PropTypes.number,
  referralRequired: PropTypes.bool,
};

const mapStateToProps = ({
  claimType,
  form: {
    claimDetailsForm: {
      values: {
        claimTypeId,
        isMultiInsurer,
        selectedPatientId = '',
        claimReason,
        isMaternity,
      } = {},
    } = {},
  },
  wallet: { balanceMap },
  user: { membersMap },
}) => ({
  claimType,
  claimTypeId,
  claimReason,
  isMaternity,
  isMultiInsurer,
  availableBalance: balanceMap[selectedPatientId],
  terminationDate: membersMap[selectedPatientId].terminationDate,
});

const enhance = compose(
  connect(mapStateToProps, { fetchWallet }),
  reduxForm({
    form: 'claimDetailsForm',
    destroyOnUnmount: false,
  }),
  injectIntl,
);

export default enhance(ClaimDetailsFormScreen);
