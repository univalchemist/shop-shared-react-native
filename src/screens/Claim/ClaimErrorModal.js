import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { Box, Button, Image, StatusPanel } from '@wrappers/components';
import { CLAIM_PATIENT_DETAILS, CLAIMS_LIST } from '@routes';
import { reset } from 'redux-form';
import { connect } from 'react-redux';
import { useIntl, useBackButtonHandler } from '@wrappers/core/hooks';
import { errorPanel } from '@images';
import { StackActions } from '@react-navigation/native';

const ClaimErrorModal = ({ reset, navigation }) => {
  const onBackButtonPressAndroid = useCallback(() => {
    navigation.dispatch(StackActions.pop(2));
    return true;
  }, [navigation]);

  useBackButtonHandler(onBackButtonPressAndroid);

  const intl = useIntl();
  return (
    <StatusPanel
      image={
        <Image
          source={errorPanel}
          maxWidth={215}
          maxHeight={215}
          resizeMode="contain"
        />
      }
      heading={intl.formatMessage({ id: 'somethingWentWrong' })}
      description={intl.formatMessage({
        id: 'claim.submitError',
      })}
      actions={
        <>
          <Box>
            <Button
              primary
              title={intl.formatMessage({
                id: 'claim.backToReviewClaim',
              })}
              onPress={() => {
                navigation.dispatch(StackActions.pop(2));
              }}
            />
          </Box>
          <Box mt={3}>
            <Button
              secondary
              title={intl.formatMessage({ id: 'claim.makeAnotherClaimText' })}
              onPress={async () => {
                const state = navigation.dangerouslyGetState();
                const firstRoute = state?.routes?.[0];

                await navigation.reset({
                  routes: [{ name: firstRoute?.name || CLAIMS_LIST }],
                });

                reset('claimDetailsForm');

                navigation.navigate(CLAIM_PATIENT_DETAILS);
              }}
            />
          </Box>
        </>
      }
    />
  );
};

ClaimErrorModal.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func,
  }),
};

export default connect(null, { reset })(ClaimErrorModal);
