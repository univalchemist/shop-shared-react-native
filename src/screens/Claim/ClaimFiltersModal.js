import React from 'react';
import { connect } from 'react-redux';
import { useFetchActions, useIntl, useTheme } from '@wrappers/core/hooks';
import { CustomMultiselectCheckBox } from '@wrappers/components/form';
import {
  ErrorPanel,
  SectionListSkeletonPlaceholder,
} from '@wrappers/components';
import {
  getClaimFilters,
  updateClaimFilters,
  getClaims,
} from '@store/claim/actions';
import { categories, logAction } from '@store/analytics/trackingActions';

export const ClaimFiltersModal = ({
  typeAndStatusFilters,
  getClaimFilters,
  membersMap,
  membersProfileOrder,
  updateClaimFilters,
  selectedClaimFilters,
  getClaims,
  navigation,
}) => {
  const theme = useTheme();
  const intl = useIntl();

  const [isLoading, isError] = useFetchActions([getClaimFilters]);

  return isLoading ? (
    <SectionListSkeletonPlaceholder count={3} />
  ) : isError ? (
    <ErrorPanel />
  ) : (
    <CustomMultiselectCheckBox
      bodyBackgroundColor={theme.colors.gray[7]}
      buttonLabel={intl.formatMessage({ id: 'showResults' })}
      clearAllButtonLabel={intl.formatMessage({ id: 'clearAll' })}
      clearAllButtonTrackingActionParams={{
        category: categories.CLAIMS,
        action: 'Clear all filters',
      }}
      submitButtonTrackingActionParams={{
        category: categories.CLAIMS,
        action: 'Filter search',
      }}
      data={composeFilterData(
        intl,
        typeAndStatusFilters,
        membersMap,
        membersProfileOrder,
      )}
      selectedValues={selectedClaimFilters}
      onSubmit={values => {
        const logActionParams = {
          category: categories.CLAIMS_FILTER,
          action: 'Filter by',
        };
        let filterTypes = [];
        if (values.length) {
          values.forEach(item => {
            let type = item.type.replace(/ /g, '_');
            type = `filter_${type}`.toLowerCase();
            filterTypes[type] = filterTypes[type] || [];
            filterTypes[type].push(item.value);
          });

          Object.keys(filterTypes).forEach(key => {
            logActionParams[key] = filterTypes[key].join(', ').toLowerCase();
          });
        }
        logAction(logActionParams);

        updateClaimFilters(values);
        getClaims();
        navigation.goBack();
      }}
      isSectionList={true}
      onChange={() => {}}
    />
  );
};

const composeFilterData = (
  intl,
  typeAndStatusFilters,
  membersMap,
  membersProfileOrder,
) => {
  const formattedTypeAndStatusFilters = transformTypeAndStatusFilters(
    intl,
    typeAndStatusFilters,
  );
  const formattedPatientFilters = transformPatientFilters(
    intl,
    membersMap,
    membersProfileOrder,
  );

  return [].concat.apply(
    [],
    [...formattedTypeAndStatusFilters, ...formattedPatientFilters],
  );
};

const transformTypeAndStatusFilters = (intl, typeAndStatusFilters) => {
  const transformedObject = Object.keys(typeAndStatusFilters).map(key => {
    return typeAndStatusFilters[key]
      .map(filter => {
        return {
          title: intl.formatMessage({ id: `claim.claimFilters.${key}` }),
          titleValue: key,
          value: filter.code,
          label: intl.formatMessage({
            id: `claim.claimFilters.${filter.code}`,
            defaultMessage: filter.text,
          }),
        };
      })
      .filter(filter => filter.value !== 'REQUEST FOR INFORMATION');
  });

  return [].concat.apply([], transformedObject);
};

const transformPatientFilters = (intl, membersMap, membersProfileOrder) => {
  return membersProfileOrder.map(memberId => {
    return {
      title: intl.formatMessage({ id: 'claim.claimFilters.patient' }),
      titleValue: 'patient',
      value: memberId,
      label: membersMap[memberId].fullName,
    };
  });
};

const mapStateToProps = ({
  claim: { filters, selectedClaimFilters },
  user: { membersMap, membersProfileOrder },
}) => {
  return {
    typeAndStatusFilters: filters,
    membersMap,
    membersProfileOrder,
    selectedClaimFilters,
  };
};

const enhance = connect(mapStateToProps, {
  getClaimFilters,
  updateClaimFilters,
  getClaims,
});

export default enhance(ClaimFiltersModal);
