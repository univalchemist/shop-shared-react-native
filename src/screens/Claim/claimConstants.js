/* istanbul ignore file */
const DEPENDENT = 'Dependant';
const PENDING = 'Pending Verification';
const APPROVED = 'Approved';
const REJECTED = 'Rejected';

const statusMap = {
  [PENDING]: 'Pending',
  [APPROVED]: 'Approved',
  [REJECTED]: 'Rejected',
};

export { DEPENDENT, PENDING, APPROVED, REJECTED, statusMap };
