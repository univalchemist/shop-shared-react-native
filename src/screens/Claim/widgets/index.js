export {
  default as ClaimPatientDetailBackButton,
  handlePress,
} from './ClaimPatientDetailBackButton';
export { default as ClaimReviewListItem } from './ClaimReviewListItem';
export { default as ClaimsListScreenHeader } from './ClaimsListScreenHeader';
