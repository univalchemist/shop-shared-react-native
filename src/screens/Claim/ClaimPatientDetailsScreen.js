import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import {
  Container,
  Box,
  TrackedButton,
  ScrollView,
  StepProgressBar,
  FieldSkeletonPlaceholder,
  ButtonSkeletonPlaceholder,
  Text,
  ErrorPanel,
} from '@wrappers/components';
import { SelectField, InputField } from '@wrappers/components/form';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { reduxForm } from 'redux-form';
import { CLAIM_PATIENT_MODAL, CLAIM_DETAILS_FORM } from '@routes';
import { Icon } from 'react-native-elements';
import { normalizeContactNumber } from '@wrappers/core/normalizers';
import {
  validateContactNumber,
  validateContactNumberLength,
} from '@wrappers/core/validations';
import { FormattedMessage } from 'react-intl';
import {
  useIntl,
  useTheme,
  useBackButtonHandler,
  useFetchActions,
} from '@wrappers/core/hooks';
import { handlePress } from '@screens/Claim';
import { fetchClaimTypes } from '@store/claimType/actions';
import { categories } from '@store/analytics/trackingActions';

const ClaimPatientDetailsSkeletonPlaceholder = () => (
  <Box pt={4} pl={4} pr={4} mb={2}>
    <FieldSkeletonPlaceholder />
    <Box mt={36}>
      <FieldSkeletonPlaceholder />
    </Box>
    <Box mt={52}>
      <ButtonSkeletonPlaceholder />
    </Box>
  </Box>
);

const ClaimPatientDetailsScreen = ({
  handleSubmit,
  reset,
  fetchClaimTypes,
  navigation,
}) => {
  const intl = useIntl();
  const theme = useTheme();
  const intlMsg = message => intl.formatMessage({ id: `claim.${message}` });
  const [isLoading, isError] = useFetchActions([fetchClaimTypes]);

  const onBackButtonPressAndroid = useCallback(() => {
    handlePress(navigation, intl, reset);
    return true;
  }, [navigation, intl, reset]);

  useBackButtonHandler(onBackButtonPressAndroid);

  let component;
  if (isLoading) {
    component = <ClaimPatientDetailsSkeletonPlaceholder />;
  } else if (isError) {
    component = <ErrorPanel />;
  } else {
    component = (
      <ScrollView>
        <StepProgressBar currentStep={1} stepCount={4} />
        <Container>
          <Box mt={2}>
            <Box>
              <SelectField
                name="patientName"
                label={intlMsg('label.patientName')}
                onPress={() => navigation.navigate(CLAIM_PATIENT_MODAL)}
                onRight={({ color }) => (
                  <Icon name="expand-more" color={color} />
                )}
              />
            </Box>
            <Box>
              <InputField
                name="contactNumber"
                keyboardType="phone-pad"
                returnKeyType="done"
                normalize={normalizeContactNumber}
                label={intlMsg('label.contactNumber')}
                hint={intlMsg('label.hint.contactNumber')}
                validate={[validateContactNumber, validateContactNumberLength]}
                customStyles={theme.customInputStyles}
              />
            </Box>

            <Box>
              <TrackedButton
                primary
                onPress={handleSubmit(() =>
                  navigation.navigate(CLAIM_DETAILS_FORM),
                )}
                title={intlMsg('addClaimDetailsButtonText')}
                actionParams={{
                  category: categories.CLAIMS_SUBMISSION,
                  action: 'Add claims details',
                }}
              />
            </Box>
          </Box>

          <Box mt={32}>
            <Text fontSize={12} lineHeight={16} letterSpacing={0.4}>
              <FormattedMessage id="claim.conditionText" />
            </Text>
          </Box>
        </Container>
      </ScrollView>
    );
  }
  return component;
};

ClaimPatientDetailsScreen.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired,
};

const enhance = compose(
  connect(
    ({ user: { userId, membersMap } }) => ({
      initialValues: {
        patientName: membersMap[userId].fullName,
        contactNumber: membersMap[userId].contactNumber,
        selectedPatientId: membersMap[userId].memberId,
      },
    }),
    { fetchClaimTypes },
  ),
  reduxForm({
    form: 'claimDetailsForm',
    destroyOnUnmount: false,
  }),
);

export { ClaimPatientDetailsSkeletonPlaceholder };
export default enhance(ClaimPatientDetailsScreen);
