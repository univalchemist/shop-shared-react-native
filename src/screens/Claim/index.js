import ClaimsListScreen from './ClaimsListScreen';
import ClaimsListScreenDependent from './ClaimsListScreenDependent';
import ClaimDetailsScreen from './ClaimDetailsScreen';

import ClaimDetailsFormScreen from './ClaimDetailsFormScreen';
import ClaimPatientDetailsScreen from './ClaimPatientDetailsScreen';
import ClaimReviewScreen from './ClaimReviewScreen';
import ClaimTypeModal from './ClaimTypeModal';
import ClaimReasonModal from './ClaimReasonModal';
import ClaimPatientModal from './ClaimPatientModal';
import ClaimSuccessModal from './ClaimSuccessModal';
import ClaimErrorModal from './ClaimErrorModal';
import ClaimTermsConditionsModal from './ClaimTermsConditionsModal';
import ClaimUploadDocumentsScreen from './ClaimUploadDocumentsScreen';
import ClaimFiltersModal from './ClaimFiltersModal';

export {
  ClaimPatientDetailBackButton,
  ClaimReviewListItem,
  ClaimsListScreenHeader,
  handlePress,
} from './widgets';

export {
  ClaimsListScreen,
  ClaimsListScreenDependent,
  ClaimDetailsScreen,
  ClaimDetailsFormScreen,
  ClaimPatientDetailsScreen,
  ClaimReviewScreen,
  ClaimTypeModal,
  ClaimReasonModal,
  ClaimPatientModal,
  ClaimSuccessModal,
  ClaimErrorModal,
  ClaimTermsConditionsModal,
  ClaimUploadDocumentsScreen,
  ClaimFiltersModal,
};
