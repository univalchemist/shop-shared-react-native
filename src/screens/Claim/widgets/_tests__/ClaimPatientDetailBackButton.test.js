import React from 'react';
import { ClaimPatientDetailBackButton } from '@screens/Claim';
import { renderForTest } from '@testUtils';
import {
  fireEvent,
  flushMicrotasksQueue,
  act,
} from 'react-native-testing-library';
import { Alert } from 'react-native';
import messages from '@messages/en-HK.json';
import navigation from '@testUtils/__mocks__/navigation';

jest.mock('react-native/Libraries/Alert/Alert', () => ({ alert: jest.fn() }));

describe('ClaimPatientDetailBackButton', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const component = renderForTest(<ClaimPatientDetailBackButton />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should display alert when icon back button is pressed', async () => {
    const { getByProps } = renderForTest(<ClaimPatientDetailBackButton />);
    const icon = getByProps({
      testID: 'iconIcon',
    });
    await fireEvent.press(icon);
    await flushMicrotasksQueue();

    expect(Alert.alert).toHaveBeenCalledTimes(1);
    expect(Alert.alert.mock.calls[0][0]).toBe(messages.leaveThisPage);
    expect(Alert.alert.mock.calls[0][1]).toBe(
      messages['claim.submitResetWarning'],
    );
    expect(Alert.alert.mock.calls[0][2][0].text).toBe(messages.stay);
    expect(Alert.alert.mock.calls[0][2][1].text).toBe(messages.leave);
  });

  it('should navigate to CLAIMS_LIST screen when alert leave button is pressed', async () => {
    const { getByProps } = renderForTest(
      <ClaimPatientDetailBackButton navigation={navigation} />,
    );
    const icon = getByProps({
      testID: 'iconIcon',
    });
    act(() => {
      fireEvent.press(icon);
    });
    act(() => {
      Alert.alert.mock.calls[0][2][1].onPress();
    });
    await flushMicrotasksQueue();

    expect(navigation.goBack).toHaveBeenCalledTimes(1);
  });

  it('should stay on current screen when alert stay button is pressed', async () => {
    const { getByProps } = renderForTest(
      <ClaimPatientDetailBackButton navigation={navigation} />,
    );
    const icon = getByProps({
      testID: 'iconIcon',
    });
    act(() => {
      fireEvent.press(icon);
    });
    await flushMicrotasksQueue();
    act(() => {
      Alert.alert.mock.calls[0][2][0].onPress();
    });

    expect(navigation.goBack).not.toHaveBeenCalled();
  });

  it('should should reset consultation type details when pressed', async () => {
    const { getByProps } = renderForTest(
      <ClaimPatientDetailBackButton navigation={navigation} />,
    );
    const icon = getByProps({
      testID: 'iconIcon',
    });
    act(() => {
      fireEvent.press(icon);
    });
    await flushMicrotasksQueue();
  });
});
