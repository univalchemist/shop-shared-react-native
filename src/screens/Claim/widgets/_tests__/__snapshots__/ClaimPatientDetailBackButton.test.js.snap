// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClaimPatientDetailBackButton should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    onInsetsChange={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              enabled={false}
              hitSlop={
                Object {
                  "right": -700,
                }
              }
              maxDeltaY={20}
              minOffsetX={5}
              onHandlerStateChange={[Function]}
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    Object {
                      "backgroundColor": "#fff",
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": -1,
                      },
                      "shadowOpacity": 0.3,
                      "shadowRadius": 5,
                      "top": 0,
                      "width": 3,
                    }
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        accessible={true}
                        focusable={true}
                        isTVSelectable={true}
                        onClick={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Object {
                            "opacity": 1,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "paddingLeft": 15,
                            }
                          }
                        >
                          <View>
                            <View
                              replaceTheme={[Function]}
                              style={
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "transparent",
                                  "justifyContent": "center",
                                }
                              }
                              theme={
                                Object {
                                  "colors": Object {
                                    "disabled": "hsl(208, 8%, 90%)",
                                    "divider": "#bcbbc1",
                                    "error": "#ff190c",
                                    "grey0": "#393e42",
                                    "grey1": "#43484d",
                                    "grey2": "#5e6977",
                                    "grey3": "#86939e",
                                    "grey4": "#bdc6cf",
                                    "grey5": "#e1e8ee",
                                    "greyOutline": "#bbb",
                                    "platform": Object {
                                      "android": Object {
                                        "error": "#f44336",
                                        "primary": "#2196f3",
                                        "secondary": "#9C27B0",
                                        "success": "#4caf50",
                                        "warning": "#ffeb3b",
                                      },
                                      "ios": Object {
                                        "error": "#ff3b30",
                                        "primary": "#007aff",
                                        "secondary": "#5856d6",
                                        "success": "#4cd964",
                                        "warning": "#ffcc00",
                                      },
                                    },
                                    "primary": "#2089dc",
                                    "searchBg": "#303337",
                                    "secondary": "#8F0CE8",
                                    "success": "#52c41a",
                                    "warning": "#faad14",
                                  },
                                }
                              }
                              underlayColor="white"
                              updateTheme={[Function]}
                            >
                              <Text
                                allowFontScaling={false}
                                style={
                                  Array [
                                    Object {
                                      "color": "#646464",
                                      "fontSize": 24,
                                    },
                                    Object {
                                      "backgroundColor": "transparent",
                                    },
                                    Object {
                                      "fontFamily": "Material Icons",
                                      "fontStyle": "normal",
                                      "fontWeight": "normal",
                                    },
                                    Object {},
                                  ]
                                }
                                testID="iconIcon"
                              >
                                î—„
                              </Text>
                            </View>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
